app/controllers/concerns/authentication.rb

```diff
@@ -52,22 +52,32 @@ def user
 
   def rbac_allowed?
     return true if ActiveModel::Type::Boolean.new.cast(Settings.disable_rbac)
-    return true if skip_rbac_for_cert_based_auth?
+    return valid_cert_auth? if identity_header.cert_based?
 
     @rbac_api ||= ::RbacApi.new(request.headers['X-RH-IDENTITY'])
     @rbac_api.check_user
   end
 
   private
 
-  def skip_rbac_for_cert_based_auth?
-    identity_header.cert_based? &&
-      ALLOWED_CERT_BASED_RBAC_ACTIONS.include?(controller: controller_name,
-                                               action: action_name)
+  def valid_cert_endpoint?
+    ALLOWED_CERT_BASED_RBAC_ACTIONS.include?(
+      controller: controller_name, action: action_name
+    )
+  end
+
+  def valid_cert_auth?
+    valid_cert_endpoint? && HostInventoryAPI.new(
+      nil, Settings.host_inventory_url, raw_identity_header
+    ).hosts.dig('results').present?
```

Comment:
If I understand this correctly, we are using Inventory API for checking the auth. Does that need to use an API that'd likely return big chunk of data to just check whether it returns something? Is there a shorter option, e.g. using lighter endpoint or `HEAD`?

How are non-200 HTTP states handled? Wouldn't it crash on 401 form Inventory with 500 on our side?

Reply:
It doesn't appear HEAD returns any useful information about the host list, unfortunately.

Reply:
Can you please elaborate why it has to request all hosts from Inventory when it does not need any of those for our auth?

Reply:
Note that this isn't actually requesting all hosts. Instead, when the cert-auth identity header of 'system' type is sent to inventory, only the hosts that have a matching `owner_id` to the `system.cn` field of the identity are returned. I agree, it's not ideal, but Inventory hasn't provided a better way of validating the identity.

Reply:
Oh, nice! :+1:  Sorry, I've haven't understood it the first time from the description.